segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/ContainerTableExtensionImpl.java

```diff
@@ -269,16 +311,49 @@ public void close() {
         return builder.getResultFutures();
     }
 
+    @SuppressWarnings("unchecked")
+    private <T, V extends Collection<T>> V translateItems(V items, SegmentProperties segmentInfo, boolean isExternal,
+                                                          BiFunction<KeyTranslator, T, T> translateFunction) {
+        if (!ContainerSortedKeyIndex.isSortedTableSegment(segmentInfo)) {
+            // Nothing to translate for non-sorted segments.
+            return items;
+        }
+
+        val t = isExternal ? SortedKeyIndexDataSource.EXTERNAL_TRANSLATOR : SortedKeyIndexDataSource.INTERNAL_TRANSLATOR;
+        return (V) items.stream().map(i -> translateItem(i, t, translateFunction)).collect(Collectors.toList());
+    }
+
+    private <T> T translateItem(T item, KeyTranslator translator, BiFunction<KeyTranslator, T, T> translateItem) {
+        return item == null ? null : translateItem.apply(translator, item);
+    }
+
     @Override
     public CompletableFuture<AsyncIterator<IteratorItem<TableKey>>> keyIterator(String segmentName, IteratorArgs args) {
-        logRequest("keyIterator", segmentName);
-        return newIterator(segmentName, args, TableBucketReader::key);
+        return this.segmentContainer.forSegment(segmentName, args.getFetchTimeout())
+                .thenComposeAsync(segment -> {
+                    if (ContainerSortedKeyIndex.isSortedTableSegment(segment.getInfo())) {
```

Comment:
If ai understand the flow right, `ContainerTableExtensionImpl` implements `ContainerTableExtension`, which implements `TableStore`. `keyIterator` and `entryIterator` are calls in the `TableStore` interface, so that's how code using table segments use the sorted table segments.

`SegmentSortedKeyIndex` is made available via `ContainerKeyIndex`, which is used in this class. This is how the `TableStore` interface connects with the `SegmentSortedKeyIndex`, is it right?

Reply:
First paragraph is 100% spot on.

Second paragraph:
- `SortedKeyIndex` is indeed exposed via `ContainerKeyIndex`, but it is only used in `ContainerTableExtensionImpl`, which, depending on whether the Segment is Sorted or not, uses it or uses the default iterator. 
- The connection is done in the `ContainerTableExtensionImpl` and again (in the back-end) in `WriterTableProcessor` that invokes its `persistUpdate` API to insert or remove keys (only when keys are inserted or removed, but not on update).

* Clarification in case I didn't explain it properly: the BTreeSet behind the SKI is only updated when a Key is inserted or removed. It is not touched when a Key is updated. Hence the frequency of writes to that data structure is a function of inserts or deletes, and not in-place updates.